### ISSUE TITLE
Allow concatenated strings to be cast as const type

### ISSUE THREAD

**Issue by Lutymane** [microsoft/TypeScript#54161]
# Suggestion

## 🔍 Search Terms

Concatenated strings const type

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## ⭐ Suggestion

Allow `as const` cast to be applied to concatenated string literals like so:

```ts
// s0 would be 'ab'
const s0 = ('a' + 'b') as const;
```

https://www.typescriptlang.org/play?#code/FAehAIGcAZwdwPYFcA2ATcAjApuA5AIaZ7ADGCAdpAC5SwC84AFIXuANT7ECU4Bk4clWoBuYKAiQAjOACWAwsTKUaUGY0Vt+glaKA

## 📃 Motivating Example

It's convenient to split a long string into multiple concatenated lines to enhance code readability, so `as const` cast should work for it as well.

## 💻 Use Cases

Currently only a single string is allowed to be cast to its literal type. So as a workaround you just sacrifice code readability and make multiple strings into one:

`'a' + 'b' => 'ab'`

**Comment by fatcerberus** [comment]
> It's convenient to split a long string into multiple concatenated lines to enhance code readability

This is why I've always wished JS supported splitting a string into multiple adjacent literals as in C, e.g. `"a" "b"` compiles down to `"ab"`

**Comment by Josh-Cena** [comment]
`"a" "b"` is so error prone—I've seen many Python programmers complaining that missing a comma in an array literal does not lead to an error but generates different behavior.

What's undesirable about this anyway?

```js
const longLiteral = "a\
b";
```

**Comment by fatcerberus** [comment]
@Josh-Cena The continuation lines can’t be indented.

**Comment by jcalz** [comment]
Duplicate of #44905

**Comment by Lutymane** [comment]
@jcalz I don't believe my issue is a duplicate, because it's specifically targeted for concatenating literal **values**, not variables with literal types. And the purpose is to provide convenience for the coding. The argument made by ahejlsberg on that issue doesn't seem to apply here as well, because strings will not change their display value, so `"a" + "b"` will **always** remain `"ab"` not something else like in case with numbers: `2 + 3` will be `5`, not `23`. So there's absolutely no need to handle any unions or other operators
